fe/fe-core/src/main/java/com/starrocks/alter/OnlineOptimizeJobV2.java

```diff
@@ -0,0 +1,806 @@
+            String partitionName = partitionNames.get(i);
+            String rewriteSql = "insert into " + dbName + "." + tableName + " TEMPORARY PARTITION ("
+                    + tmpPartitionName + ") select " + Joiner.on(", ").join(tableColumnNames)
+                    + " from " + dbName + "." + tableName + " partition (" + partitionName + ")";
+            String taskName = getName() + "_" + tmpPartitionName;
+            OptimizeTask rewriteTask = TaskBuilder.buildOptimizeTask(taskName, properties, rewriteSql, dbName);
+            rewriteTask.setPartitionName(partitionName);
+            rewriteTask.setTempPartitionName(tmpPartitionName);
+            rewriteTasks.add(rewriteTask);
+        }
+
+        this.jobState = JobState.RUNNING;
+        span.addEvent("setRunning");
+
+        // DO NOT write edit log here, tasks will be send again if FE restart or master changed.
+        LOG.info("transfer optimize job {} state to {}", jobId, this.jobState);
+    }
+
+    private void enableDoubleWritePartition(Database db, OlapTable tbl, String sourcePartitionName, String tmpPartitionName) {
+        Locker locker = new Locker();
+        locker.lockDatabase(db, LockType.WRITE);
+        try {
+            Preconditions.checkState(tbl.getState() == OlapTableState.SCHEMA_CHANGE);
+            tbl.addDoubleWritePartition(sourcePartitionName, tmpPartitionName);
+            LOG.info("job {} add double write partition {} to {}", jobId, tmpPartitionName, sourcePartitionName);
+        } finally {
+            locker.unLockDatabase(db, LockType.WRITE);
+        }
+    }
+
+    private void disableDoubleWritePartition(Database db, OlapTable tbl) {
+        Locker locker = new Locker();
+        locker.lockDatabase(db, LockType.WRITE);
+        try {
+            Preconditions.checkState(tbl.getState() == OlapTableState.SCHEMA_CHANGE);
+            tbl.clearDoubleWritePartition();
+            LOG.info("job {} clear double write partitions", jobId);
+        } finally {
+            locker.unLockDatabase(db, LockType.WRITE);
+        }
+    }
+
+    /**
+     * runRunningJob()
+     * 1. Wait insert into tasks to be finished.
+     * 2. Replace partitions with temp partitions.
+     * 3. Set job'state as FINISHED.
+     */
+    @Override
+    protected void runRunningJob() throws AlterCancelException {
+        Preconditions.checkState(jobState == JobState.RUNNING, jobState);
+
+        Database db = GlobalStateMgr.getCurrentState().getDb(dbId);
+        if (db == null) {
+            throw new AlterCancelException("Database " + dbId + " does not exist");
+        }
+
+        OlapTable tbl;
+        Locker locker = new Locker();
+        locker.lockDatabase(db, LockType.READ);
+        try {
+            tbl = checkAndGetTable(db, tableId);
+        } finally {
+            locker.unLockDatabase(db, LockType.READ);
+        }
+
+        int progress = 0;
+
+        for (OptimizeTask rewriteTask : rewriteTasks) {
```

Comment:
rewrite tasks are expected to be executed serially？

Reply:
yes, minilize the double write impact.